File: 1046-last-stone-weight/1046-last-stone-weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        
        priority_queue<int> pq;
        
        for(auto it:stones){
            pq.push(it);
        }
        
        while(pq.size()>1){
            int stone1 = pq.top();
            pq.pop();
            int stone2 = pq.top();
            pq.pop();
            
            if(stone1!=stone2){
                pq.push(stone1-stone2);  // as we remove stone1 from heap first it will be larger
            }
        }
        
        
        return pq.empty() ? 0:pq.top();
        
        
    }
};

File: 1046-last-stone-weight/1046-last-stone-weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        
        
        if len(stones)==1:
            return stones[0]
        
        for i in range(len(stones)):
            stones[i]*=-1
        
        heapq.heapify(stones)
        
        while len(stones)>1:
            
            first = heapq.heappop(stones)
            
            second = heapq.heappop(stones)
            
            if first!=second:
                heapq.heappush(stones,first-second)
            
        return -heapq.heappop(stones) if stones else 0
        
        